File: practices/practice4.1/test/src/ConcretePizzaBuilderTest.cpp
#include "MyTest.hpp"
#include "ConcretePizzaBuilder.hpp"

void ConcretePizzaBuilderTest(UnitTests& testSuite) {
    testSuite.addTest("ConcretePizzaBuilderAddIngredient",
    [](){
        // Build:
        std::map<std::string, std::pair<double, bool>> ingredients = {
                {"Cheese", {1.0, true}},
                {"Pepperoni", {1.5, false}},
                {"Olives", {0.75, false}}
        };

        // Operate:
        std::unique_ptr<PizzaBuilder> builder = std::make_unique<ConcretePizzaBuilder>(ingredients);
        builder->addIngredient("Pepperoni", 2);

        // Check:
        ASSERT_EQ(builder->getPizza()->getIngredientCount("Pepperoni"), 2)
    });

    testSuite.addTest("ConcretePizzaBuilderRemoveIngredient",
    [](){
        // Build:
        std::map<std::string, std::pair<double, bool>> ingredients = {
                {"Cheese", {1.0, true}},
                {"Pepperoni", {1.5, false}},
                {"Olives", {0.75, false}}
        };

        // Operate:
        std::unique_ptr<PizzaBuilder> builder = std::make_unique<ConcretePizzaBuilder>(ingredients);
        builder->addIngredient("Olives", 2);
        builder->removeIngredient("Olives", 1);

        // Check:
        ASSERT_EQ(builder->getPizza()->getIngredientCount("Olives"), 1)
    });
}

File: homeworks/homework4.1/src/test/java/com/pizza/ConcretePizzaBuilderTest.java
package com.pizza;

import static org.junit.Assert.*;
import org.junit.Test;
import java.util.Map;

public class ConcretePizzaBuilderTest {

    @Test
    public void testAddIngredient() {
        Map<String, Ingredient> ingredients = Map.of(
                "Cheese", new Ingredient(1.0, true, 1),
                "Pepperoni", new Ingredient(1.5, false, 1),
                "Olives", new Ingredient(0.75, false, 1)
        );

        ConcretePizzaBuilder builder = new ConcretePizzaBuilder(ingredients);
        builder.addIngredient("Pepperoni", 2);

        assertEquals(2, builder.getPizza().getIngredientCount("Pepperoni"));
    }

    @Test
    public void testRemoveIngredient() {
        Map<String, Ingredient> ingredients = Map.of(
                "Cheese", new Ingredient(1.0, true, 1),
                "Pepperoni", new Ingredient(1.5, false, 1),
                "Olives", new Ingredient(0.75, false, 1)
        );

        ConcretePizzaBuilder builder = new ConcretePizzaBuilder(ingredients);
        builder.addIngredient("Olives", 2);
        builder.removeIngredient("Olives", 1);

        assertEquals(1, builder.getPizza().getIngredientCount("Olives"));
    }
}
